core/services/periodicbackup/backup.go

```diff
@@ -0,0 +1,139 @@
+package periodicbackup
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/chainlink/core/logger"
+)
+
+var (
+	fileName           = "db_backup.tar.gz"
+	minBackupFrequency = time.Minute
+)
+
+type backupResult struct {
+	size int64
+	path string
+}
+
+type (
+	DatabaseBackup interface {
+		Start() error
+		Close() error
+		RunBackupGracefully()
+	}
+
+	databaseBackup struct {
+		logger          *logger.Logger
+		databaseURL     url.URL
+		frequency       time.Duration
+		outputParentDir string
+		done            chan bool
+	}
+)
+
+func NewDatabaseBackup(frequency time.Duration, databaseURL url.URL, outputParentDir string, logger *logger.Logger) DatabaseBackup {
+	return &databaseBackup{
+		logger,
+		databaseURL,
+		frequency,
+		outputParentDir,
+		make(chan bool),
+	}
+}
+
+func (backup databaseBackup) Start() error {
+
+	ticker := time.NewTicker(backup.frequency)
+
+	go func() {
+		for {
+			select {
+			case <-backup.done:
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				if backup.frequencyIsTooSmall() {
+					logger.Errorf("Database backup frequency (%s=%v) is too small. Please set it to at least %s", "DATABASE_BACKUP_FREQUENCY", backup.frequency, minBackupFrequency)
+					continue
+				}
+				backup.RunBackupGracefully()
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (backup databaseBackup) Close() error {
+	backup.done <- true
+	return nil
+}
+
+func (backup *databaseBackup) frequencyIsTooSmall() bool {
+	return backup.frequency < minBackupFrequency
+}
+
+func (backup *databaseBackup) RunBackupGracefully() {
+	backup.logger.Info("DatabaseBackup: Starting database backup...")
+	startAt := time.Now()
+	result, err := backup.runBackup()
+	duration := time.Since(startAt)
+	if err != nil {
+		backup.logger.Errorf("DatabaseBackup: Failed after %s with: %v", duration, err)
```

Comment:
Can we use Errorw and Infow for structured logging?

Reply:
yup